other_packages/python/tiledbsoma_ml/src/tiledbsoma_ml/pytorch.py

```diff
@@ -0,0 +1,1173 @@
+        else:
+            # else, it is caller responsibility to open/close the experiment
+            exp_cm = contextlib.nullcontext(exp)
+
+        with exp_cm as exp:
+            with exp.axis_query(
+                measurement_name=self.measurement_name,
+                obs_query=self.obs_query,
+                var_query=self.var_query,
+            ) as query:
+                self._obs_joinids = query.obs_joinids().to_numpy()
+                self._var_joinids = query.var_joinids().to_numpy()
+
+        self._create_obs_joinids_partition()
+        self._initialized = True
+
+    def __iter__(self) -> Iterator[XObsDatum]:
+        """Create iterator over query.
+
+        Returns:
+            ``iterator``
+
+        Lifecycle:
+            experimental
+        """
+        with self.experiment_locator.open_experiment() as exp:
+            self._init_once(exp)
+            X = exp.ms[self.measurement_name].X[self.layer_name]
+            if not isinstance(X, soma.SparseNDArray):
+                raise NotImplementedError(
+                    "ExperimentAxisQueryIterDataPipe only supported on X layers which are of type SparseNDArray"
+                )
+
+            assert self._obs_joinids_partition is not None
+            obs_joinid_iter = iter(self._obs_joinids_partition)
+            _mini_batch_iter = self._mini_batch_iter(exp.obs, X, obs_joinid_iter)
+            if self.use_eager_fetch:
+                _mini_batch_iter = _EagerIterator(
+                    _mini_batch_iter, pool=exp.context.threadpool
+                )
+
+            yield from _mini_batch_iter
+
+    def __len__(self) -> int:
+        """Return approximate number of batches this iterable will produce. If run in the context of :class:`torch.distributed` or
+        as a multi-process loader (i.e., :class:`torch.utils.data.DataLoader` instantiated with num_workers > 0), the obs (cell)
+        count will reflect the size of the partition of the data assigned to the active process.
+
+        See import caveats in the PyTorch
+        [:class:`torch.utils.data.DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader)
+        domentation regarding ``len(dataloader)``, which also apply to this class.
+
+        Returns:
+            An ``int``.
+
+        Lifecycle:
+            experimental
+        """
+        return self.shape[0]
+
+    @property
+    def shape(self) -> Tuple[int, int]:
+        """Get the approximate shape of the data that will be returned by this :class:`tiledbsoma_ml.ExperimentAxisQueryIterable`.
+        This is the number of obs (cell) and var (feature) counts in the returned data. If used in multiprocessing mode
+        (i.e. :class:`torch.utils.data.DataLoader` instantiated with num_workers > 0), the obs (cell) count will reflect
+        the size of the partition of the data assigned to the active process.
+
+        Returns:
+            A 2-tuple of ``int``s, for obs and var counts, respectively.
```

Comment:
```suggestion
            A tuple of two ``int`` values: number of obs, number of vars.
```

Totally subjective, but `2-tuple` tripped me up a bit. Def ignore this suggestion if that's a standard term.